initial version of a Onkyo binding

pail23 opened the pull request:
This is a port of the openhab1 Onkyo binding to openhab2. Not all features of the openhab1 binding are available yet.

Current features include:
- Power on/off
- Volume 
- Select input
- Mute
- Player, Album, Artist and Title for the network or USB input of the Onkyo receiver
- Discovery of Onkyo Receivers

Uses the protocol implementation of the original openhab1 Onkyo binding,

kaikreuzer commented on addons/binding/org.openhab.binding.onkyo/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,168 @@
+            <parameter name="ipAddress" type="text" required="true">
+                <label>IP or Host Name</label>
+                <description>The IP or host name of the Onkyo Receiver
+                </description>
```
remove the line break

kaikreuzer commented on addons/binding/org.openhab.binding.onkyo/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,168 @@
+        
+        <config-description>
+            <parameter name="ipAddress" type="text" required="true">
+                <label>IP or Host Name</label>
```
add the line
```
<context>network-address</context>
```

kaikreuzer replied:
We usually call this label "Network Address"

kaikreuzer commented on addons/binding/org.openhab.binding.onkyo/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,168 @@
+            </parameter>
+            <parameter name="port" type="integer" required="true"
+                min="1" max="65335">
+                <label>Onkyo Receiver Port</label>
```
shorten to "Port"

kaikreuzer commented on addons/binding/org.openhab.binding.onkyo/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,168 @@
+        
+        <config-description>
+            <parameter name="ipAddress" type="text" required="true">
+                <label>IP or Host Name</label>
```
same comments as above

kaikreuzer commented on addons/binding/org.openhab.binding.onkyo/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,168 @@
+        <state min="0" max="100" step="1" pattern="%d %%">
+        </state>
+    </channel-type>
+    <channel-type id="next">
```
I think this channel can be removed as the control channel supports the same command.

kaikreuzer commented on addons/binding/org.openhab.binding.onkyo/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,168 @@
+        <label>Next</label>
+        <description>Send a next command to the player</description>
+    </channel-type>
+    <channel-type id="prev">
```
I think this channel can be removed as the control channel supports the same command.

tavalin replied:
I'd echo @kaikreuzer's comments.  I just had a play with the UI and these channels have the same outcome as the previous/next control channels so they're seem redundant.

kaikreuzer commented on addons/binding/org.openhab.binding.onkyo/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,168 @@
+    <!-- Onkyo variables -->
+    <channel-type id="currentPlayingTime">
+        <item-type>String</item-type>
+        <label>Current Playing Time</label>
```
any shorter label possible? "Playing Time" is probably enough.

kaikreuzer commented on addons/binding/org.openhab.binding.onkyo/META-INF/MANIFEST.MF:
```diff
@@ -0,0 +1,24 @@
+ org.jupnp.model.types,
+ org.osgi.service.component,
+ org.slf4j
+Service-Component: OSGI-INF/*
```
please chance to OSGI-INF/*.xml

kaikreuzer commented on addons/binding/org.openhab.binding.onkyo/README.md:
```diff
@@ -0,0 +1,78 @@
+This binding integrates the Onkyo AV receivers.
+
+## Introduction
+Binding should be compatible with Onkyo AV receivers which support ISCP (Integra Serial Control Protocol) over Ethernet (eISCP).
```
Please add an empty line after each header as otherwise layout on docs.openhab.org will break.

kaikreuzer commented on addons/binding/org.openhab.binding.onkyo/README.md:
````diff
@@ -0,0 +1,78 @@
+org.openhab.onkyo:enableAutoDiscovery=false
+```
+
+This configuration parameter only controls the Onkyo AVR auto-discovery process, not the OpenHAB auto-discovery. Moreover, if the OpenHAB auto-discovery is disabled, the Onkyo AVR auto-discovery is disabled too.
````
note that openHAB is spelled with a small "o".

kaikreuzer commented on addons/binding/org.openhab.binding.onkyo/README.md:
```diff
@@ -0,0 +1,78 @@
+| title                   | String       | Title of the current song (available if playing from Network or USB)|
+| album                   | String       | Album name of the current song (available if playing from Network or USB)|
+| artist                  | String       | Artist name of the current song (available if playing from Network or USB)|
+| currentPlayingTime                  | String       | Current playing time of the current song (available if playing from Network or USB)|
```
there are a few spaces too much

kaikreuzer commented on addons/binding/org.openhab.binding.onkyo/src/main/java/org/openhab/binding/onkyo/handler/OnkyoHandler.java:
```diff
@@ -0,0 +1,344 @@
+
+    public OnkyoHandler(Thing thing) {
+        super(thing);
+        String host = (String) this.getConfig().get(HOST_PARAMETER);
```
All this code should be in initialize() - the constructor should stay empty.

tavalin commented on addons/binding/org.openhab.binding.onkyo/README.md:
```diff
@@ -0,0 +1,78 @@
+## Discovery
+This binding can discover the supported Onkyo AV Receivers. At the moment only the following models are supported:
+
+* TX-NR535
```
Is there some sort of constraint such that only the TX-NR535 is supported? Or is that the only one this has been tested with?

pail23 replied:
I have only tested it with the TX-NR535. Probably other models will work as well since the communication part of this binding is based on the OH1 Onkyo binding.

tavalin replied:
I just tested it on my TX-NR616 and it works. Due to your supported model check it looks like it's a code change [here](https://github.com/openhab/openhab2-addons/pull/768/files#diff-6a4118c696d2ddc35356333392a8f11bR27) every time a model is verified as working.

kaikreuzer commented on addons/binding/org.openhab.binding.onkyo/src/main/java/org/openhab/binding/onkyo/handler/OnkyoHandler.java:
```diff
@@ -0,0 +1,344 @@
+                }
+                break;
+            default:
+                logger.debug("Received unknown channel {}", channelUID.getIdWithoutGroup());
```
You probably mean "unknown command".
Note that you should also support the REFRESH command.

kaikreuzer commented on addons/binding/org.openhab.binding.onkyo/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,168 @@
+    </thing-type>
+
+    <!-- Onkyo unknown Thing Type -->
+    <thing-type id="onkyoUnsupported">
```
you can add a `listed="false"` here, which will suppress this entry when manually adding a thing.

kaikreuzer commented on addons/binding/org.openhab.binding.onkyo/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,168 @@
+    </channel-type>
+    <channel-type id="input">
+        <item-type>Number</item-type>
+        <label>Input source</label>
```
change to "Input Source" to be in line with the capitalization of the other labels

kaikreuzer commented:
Hi @pail23, many thanks for this contribution and sorry that it took me soooo long to review, although there are really only some minor remarks.

> Not all features of the openhab1 binding are available yet.

Could you elaborate on what is missing? Do you plan to get it on par before we merge it?

tavalin commented:
@pail23 is there a reason why approving the new discovered Thing turns the AVR off? That was quite annoying.

pail23 commented:
@tavalin: I think I have fixed this issue. Could you please test again and provide feedback?
@kaikreuzer : Thanks a lot for the review. I have adjusted the code according your comments.

The following features are missing at the moment:
- Zone 2 and Zone 3: I dont use Zone2/3. Therefore I would not be able to test it
- Video Wide setting: My receiver seams not to support this setting. 
- The OH1 bindung allows to send raw commands to the receiver. I don't know if this really makes sense in OH2 and how to map this to the channel concept.

I propose the merge the binding with the current feature set. If somebody else has a setup with zone 2/3 or a receiver which supports Video Wide setting, the implementation can still be extended.

tavalin commented:
@pail23 Tested the latest commits and restarted OH twice, no change in AVR states this time :+1:

kaikreuzer commented:
Many thanks @pail23!
